File: lib/resource-taxonomy.ts
// Shared taxonomy for persona / industry tagging across Agents, Ops Boards, and KBs.
//
// Source of truth for TypeScript code: ./resource-taxonomy.json.
// Python equivalent (Numa KB lambda): lib/kb-core/kb_core/resource_taxonomy.py
//   — keep that file in sync if the values change.
import { z } from 'zod';
import taxonomy from './resource-taxonomy.json';

export const PERSONAS: readonly string[] = taxonomy.personas;
export const INDUSTRIES: readonly string[] = taxonomy.industries;

export type Persona = string;
export type Industry = string;

// Strict Zod schemas — exact (case-sensitive) match against the taxonomy.
export const personaSchema = z.enum(taxonomy.personas as [string, ...string[]]);
export const industrySchema = z.enum(taxonomy.industries as [string, ...string[]]);
export const personasArraySchema = z.array(personaSchema);
export const industriesArraySchema = z.array(industrySchema);

export type NormaliseResult = { values: string[]; invalid: string[] };

const normaliseAgainst = (input: unknown, allowed: readonly string[]): NormaliseResult => {
  if (!Array.isArray(input)) return { values: [], invalid: [] };
  const lowerToCanonical = new Map(allowed.map((v) => [v.toLowerCase(), v]));
  const seen = new Set<string>();
  const values: string[] = [];
  const invalid: string[] = [];
  for (const v of input) {
    if (typeof v !== 'string') continue;
    const trimmed = v.trim();
    if (!trimmed) continue;
    const canonical = lowerToCanonical.get(trimmed.toLowerCase());
    if (!canonical) {
      invalid.push(trimmed);
      continue;
    }
    if (seen.has(canonical)) continue;
    seen.add(canonical);
    values.push(canonical);
  }
  return { values, invalid };
};

// Lenient on whitespace and case, strict on membership.
// Callers decide whether to reject when invalid.length > 0.
export const normalisePersonas = (input: unknown): NormaliseResult => normaliseAgainst(input, PERSONAS);
export const normaliseIndustries = (input: unknown): NormaliseResult => normaliseAgainst(input, INDUSTRIES);

File: lib/kb-core/kb_core/resource_taxonomy.py
"""Shared taxonomy for persona / industry tagging on Knowledge Bases.

TypeScript equivalent: lib/resource-taxonomy.ts + lib/resource-taxonomy.json
Keep these files in sync if values change.
"""

from typing import List, Tuple

PERSONAS: Tuple[str, ...] = ("CEO", "Finance", "HR", "Operations", "Commercial")
INDUSTRIES: Tuple[str, ...] = (
    "Manufacturing",
    "Construction",
    "Engineering",
    "Professional Services",
    "Franchise",
)


def _normalise_against(
    values: object, allowed: Tuple[str, ...]
) -> Tuple[List[str], List[str]]:
    """Trim, dedupe case-insensitively, return canonical case.

    Returns (valid_values, invalid_values). `valid_values` are taxonomy
    members in canonical case; `invalid_values` are trimmed input strings
    that did not match any taxonomy entry.
    """
    if not isinstance(values, list):
        return [], []
    lower_to_canonical = {v.lower(): v for v in allowed}
    seen: set[str] = set()
    valid: List[str] = []
    invalid: List[str] = []
    for v in values:
        if not isinstance(v, str):
            continue
        trimmed = v.strip()
        if not trimmed:
            continue
        canonical = lower_to_canonical.get(trimmed.lower())
        if canonical is None:
            invalid.append(trimmed)
            continue
        if canonical in seen:
            continue
        seen.add(canonical)
        valid.append(canonical)
    return valid, invalid


def normalise_personas(values: object) -> Tuple[List[str], List[str]]:
    return _normalise_against(values, PERSONAS)


def normalise_industries(values: object) -> Tuple[List[str], List[str]]:
    return _normalise_against(values, INDUSTRIES)
